Move last post updating into CategoryModel

initvector opened the pull request:
Per the original issue, this update:

1. Stores `LastCategoryID`, `CountAllDiscussions` and `CountAllComments` fields in the category table.
1. Refactors count and latest post updating into `CategoryModel`.
1. Increments the CountAll columns, when posting/deleting discussions/comments.
1. Sets the `LastCategoryID` to the current category ID when posting a discussion or comment.

Closes #3902

#### Notes
1. This was a little hairy to sort out, so it warrants some additional in-depth testing by the reviewer.
1. ~~You should run /dba/counts before starting, particularly the `Recalculate Category.CountAllDiscussions` and `Recalculate Category.CountAllComments` jobs.~~
1. This PR should be merged into the `fix/category-optimization` branch.

linc commented:
Why is this reassigned to @DaazKu with month-stale feedback still waiting to be addressed?

initvector commented:
She's no longer stale. Fresh as a daisy, she is.

Feedback has been addressed and branches (`fix/category-optimization`, `fix/categorymodel-counts-lastpost`) have been rebased as they were several weeks old. Please make sure you remove any local copies of these branches and pull from origin.

DaazKu commented:
This PRs wants to merge in fix/category-optimization. Shouldn't it be remade to merge in master?

DaazKu commented:
Replaced by https://github.com/vanilla/vanilla/pull/5388

initvector commented:
I've updated the PR to automatically build the necessary counts when the columns are added.

In addition, I was running into an issue where counts were sometimes failing (MySQL's `sum` returning a `null` value), so I added a call to `coalesce` on the count fields.

DaazKu commented:
Nice work! This PR works great!

<img width="889" alt="screen shot 2017-03-30 at 4 07 55 pm" src="https://cloud.githubusercontent.com/assets/2412909/24524111/7a7d070e-1563-11e7-9dbd-97a21216a180.png">

It was 0 and 0 before!

DaazKu commented on applications/vanilla/models/class.categorymodel.php:
```diff
@@ -1035,6 +1039,155 @@ private function gatherLastIDs($categoryTree, &$result = null) {
+     */
+    public function incrementLastDiscussion($discussion) {
+        // Lookup the discussion record, if necessary. We need at least a discussion to continue.
+        if (is_numeric($discussion)) {
```
`is_numeric(9.1) === true`
`is_numeric('0xf4c3b00c') === true` prior to php7.0

Use `filter_var($value, FILTER_VALIDATE_INT) !== false` or `ctype_digit((string)$value)`

DaazKu commented on applications/vanilla/models/class.discussionmodel.php:
```diff
@@ -2267,38 +2274,17 @@ public function updateDiscussionCount($CategoryID, $Discussion = false) {
-            ->where('CategoryID', val('CategoryID', $Discussion))
-            ->put();
+    public function incrementNewDiscussion($discussion) {
+        trigger_error(
```
We have the `deprecated()` function just for that!

DaazKu commented on applications/vanilla/settings/structure.php:
```diff
@@ -423,6 +428,15 @@
         ->put();
 }
 
+if (!$captureOnly) {
```
Maybe we could put the exist check in here since they are only used if captureOnly is false?

initvector replied:
If the is-existing check is done this far down, the columns have already been created if they didn't exist in the first place. It'll always return `true`.

DaazKu replied:
Right. Good then! :)
